wgpu-core/src/device/mod.rs

```diff
@@ -3593,6 +3599,141 @@ impl<G: GlobalIdentityHandlerFactory> Global<G> {
+                && !device
+                    .features
+                    .contains(wgt::Features::PIPELINE_STATISTICS_QUERY)
+            {
+                break resource::CreateQuerySetError::MissingFeature(
+                    wgt::Features::PIPELINE_STATISTICS_QUERY,
+                );
+            }
+
+            if desc.count == 0 {
+                break resource::CreateQuerySetError::ZeroCount;
+            }
+
+            if desc.count == wgt::QUERY_SET_MAX_QUERIES {
+                break resource::CreateQuerySetError::TooManyQueries {
+                    count: desc.count,
+                    maximum: wgt::QUERY_SET_MAX_QUERIES,
+                };
+            }
+
+            let query_set = {
+                let (hal_type, elements) = conv::map_query_type(&desc.ty);
+
+                resource::QuerySet {
+                    raw: unsafe { device.raw.create_query_pool(hal_type, desc.count).unwrap() },
+                    device_id: Stored {
+                        value: id::Valid(device_id),
+                        ref_count: device.life_guard.add_ref(),
+                    },
+                    life_guard: LifeGuard::new(""),
+                    desc: desc.clone(),
+                    elements,
+                }
+            };
+
+            let ref_count = query_set.life_guard.add_ref();
+
+            let id = hub
+                .query_sets
+                .register_identity(id_in, query_set, &mut token);
+            // #[cfg(feature = "trace")]
+            // match device.trace {
+            //     Some(ref trace) => trace.lock().add(trace::Action::CreateQuerySet {
+            //         id,
+            //         desc: trace::QuerySetDescriptor {
+            //             type_: trace::QueryType::new(&desc.type_),
+            //             count: desc.count,
+            //         },
+            //     }),
+            //     None => (),
+            // };
+
+            device
+                .trackers
+                .lock()
+                .query_sets
+                .init(id, ref_count, PhantomData)
+                .unwrap();
+
+            return (id.0, None);
+        };
+
+        let id = B::hub(self)
+            .query_sets
+            .register_error(id_in, "", &mut token);
+        (id, Some(error))
+    }
+
+    pub fn device_get_timestamp_period<B: GfxBackend>(
+        &self,
+        device_id: id::DeviceId,
+    ) -> Result<f32, DeviceError> {
+        span!(_guard, INFO, "Device::get_timestamp_period");
+
+        let hub = B::hub(self);
+        let mut token = Token::root();
+        let (device_guard, _) = hub.devices.read(&mut token);
+
+        let device = match device_guard.get(device_id) {
+            Ok(device) => device,
+            Err(_) => return Err(DeviceError::Invalid),
+        };
+
+        Ok(device.private_features.timestamp_period)
+    }
+
+    pub fn query_set_drop<B: GfxBackend>(&self, query_set_id: id::QuerySetId) {
```

Comment:
I think a few cleanup stages might be missing. You can follow the bind groups for example to see where to put the logic:
  - updating the submission index of the query sets upon submission
  - when triaging suspected resources, checking for the refcount and moving to the free resources, or associated submissions
  - when considering a submission finished, moving the resources in to free resources
  - when going through the free resources, actually deleting the sets
  - when encoding any operations involving the sets, add them to the tracked list of the command buffer (I commented about it separately)